src/window/testing_clock.rs

```diff
@@ -5,118 +5,48 @@ use pyo3::{exceptions::PyValueError, prelude::*};
+use crate::pyo3_extensions::{TdPyAny, TdPyCallable};
 use crate::recovery::StateBytes;
 
-/// Encapsulates and allows modifying the "now" when using
-/// `bytewax.window.TestingClockConfig`. You only want to use this for
-/// unit testing.
-///
-/// Args:
-///
-///   init_datetime (datetime.datetime): Initial "now".
-///
-/// Returns:
-///
-///   Testing clock object. Pass this as the `clock` parameter to
-///   `bytewax.window.TestingClockConfig`.
-#[pyclass(module = "bytewax.testing", name = "TestingClock")]
-#[pyo3(text_signature = "(init_datetime)")]
-#[derive(Clone)]
-pub(crate) struct PyTestingClock {
-    /// Modify this to change the current "now".
-    #[pyo3(get, set)]
-    pub(crate) now: DateTime<Utc>,
-}
-
-impl PyTestingClock {
-    /// Create an "empty" [`Self`] just for use in `__getnewargs__`.
-    #[allow(dead_code)]
-    pub(crate) fn pickle_new(py: Python) -> Py<Self> {
-        PyCell::new(
-            py,
-            PyTestingClock {
-                now: DateTime::<Utc>::MIN_UTC,
-            },
-        )
-        .unwrap()
-        .into()
-    }
-}
-
-#[pymethods]
-impl PyTestingClock {
-    /// Tell pytest to ignore this class, even though it starts with
-    /// the name "Test".
-    #[allow(non_upper_case_globals)]
-    #[classattr]
-    const __test__: bool = false;
-
-    #[new]
-    #[args(init_datetime)]
-    fn new(init_datetime: DateTime<Utc>) -> Self {
-        Self { now: init_datetime }
-    }
-
-    /// Pickle as a tuple.
-    fn __getstate__(&self) -> (&str, DateTime<Utc>) {
-        ("TestingClock", self.now)
-    }
-
-    /// Egregious hack see [`SqliteRecoveryConfig::__getnewargs__`].
-    fn __getnewargs__(&self) -> (DateTime<Utc>,) {
-        (DateTime::<Utc>::MIN_UTC,)
-    }
-
-    /// Unpickle from tuple of arguments.
-    fn __setstate__(&mut self, state: &PyAny) -> PyResult<()> {
-        if let Ok(("TestingClock", now)) = state.extract() {
-            self.now = now;
-            Ok(())
-        } else {
-            Err(PyValueError::new_err(format!(
-                "bad pickle contents for TestingClock: {state:?}"
-            )))
-        }
-    }
-}
-
-/// Use to simulate system time in unit tests. You only want to use
-/// this for unit testing.
-///
-/// You should use this with
-/// `bytewax.inputs.TestingBuilderInputConfig` and a generator which
-/// modifies the `TestingClock` provided.
+/// In tests, we only allow the use of event time, to avoid
+/// either mocking or relying on system time.
 ///
 /// If the dataflow has no more input, all windows are closed.
-///
-/// The watermark uses the most recent "now".
+/// If wait_until_end is False, windows are closed as soon as an event
+/// of a later window is received.
 ///
 /// Args:
 ///
-///   clock (TestingClock): Query this `TestingClock` for the current
-///       "now".
+///  dt_getter: function used to retrieve a UTC datetime from each event.
+///
+///  wait_until_end: wether to wait for all data to be collected, or
```

Comment:
When I read this comment, I wonder how we know "all" data has been collected.

I'm also wondering if it makes sense to default this to `false` and only explicitly set it for cases where we want to allow infinite lateness, since it seems like default behavior is "close the window when i say the window is closed (i.e. after 10 min).

Also, maybe that points to renaming this, since it's about waiting for all data, or allowing a wide wait band for lateness.